packages/snaps-rpc-methods/src/permitted/getState.ts

```diff
@@ -0,0 +1,189 @@
+
+import { manageStateBuilder } from '../restricted/manageState';
+import type { MethodHooksObject } from '../utils';
+import { StateKeyStruct } from '../utils';
+
+const hookNames: MethodHooksObject<GetStateHooks> = {
+  hasPermission: true,
+  getSnapState: true,
+  getUnlockPromise: true,
+};
+
+/**
+ * `snap_getState` gets the state of the Snap.
+ */
+export const getStateHandler: PermittedHandlerExport<
+  GetStateHooks,
+  GetStateParameters,
+  GetStateResult
+> = {
+  methodNames: ['snap_getState'],
+  implementation: getStateImplementation,
+  hookNames,
+};
+
+export type GetStateHooks = {
+  /**
+   * Check if the requesting origin has a given permission.
+   *
+   * @param permissionName - The name of the permission to check.
+   * @returns Whether the origin has the permission.
+   */
+  hasPermission: (permissionName: string) => boolean;
+
+  /**
+   * Get the state of the requesting Snap.
+   *
+   * @returns The current state of the Snap.
+   */
+  getSnapState: (
+    snapId: string,
+    encrypted: boolean,
+  ) => Promise<Record<string, Json>>;
+
+  /**
+   * Wait for the extension to be unlocked.
+   *
+   * @returns A promise that resolves once the extension is unlocked.
+   */
+  getUnlockPromise: (shouldShowUnlockRequest: boolean) => Promise<void>;
+};
+
+const GetStateParametersStruct = object({
+  key: optional(StateKeyStruct),
+  encrypted: optional(boolean()),
+});
+
+export type GetStateParameters = InferMatching<
+  typeof GetStateParametersStruct,
+  GetStateParams
+>;
+
+/**
+ * The `snap_getState` method implementation.
+ *
+ * @param request - The JSON-RPC request object.
+ * @param response - The JSON-RPC response object.
+ * @param _next - The `json-rpc-engine` "next" callback. Not used by this
+ * function.
+ * @param end - The `json-rpc-engine` "end" callback.
+ * @param hooks - The RPC method hooks.
+ * @param hooks.hasPermission - Check whether a given origin has a given
+ * permission.
+ * @param hooks.getSnapState - Get the state of the requesting Snap.
+ * @param hooks.getUnlockPromise - Wait for the extension to be unlocked.
+ * @returns Nothing.
+ */
+async function getStateImplementation(
+  request: JsonRpcRequest<GetStateParameters>,
+  response: PendingJsonRpcResponse<GetStateResult>,
+  _next: unknown,
+  end: JsonRpcEngineEndCallback,
+  { hasPermission, getSnapState, getUnlockPromise }: GetStateHooks,
+): Promise<void> {
+  const { params } = request;
+
+  if (!hasPermission(manageStateBuilder.targetName)) {
+    return end(providerErrors.unauthorized());
+  }
+
+  try {
+    const validatedParams = getValidatedParams(params);
+    const { key, encrypted = true } = validatedParams;
+
+    if (encrypted) {
+      await getUnlockPromise(true);
+    }
+
+    // We expect the MM middleware stack to always add the origin to requests
+    const { origin } = request as JsonRpcRequest & { origin: string };
+    const state = await getSnapState(origin, encrypted);
```

Comment:
Might be preferable for the permitted implementation to simply bind all the hooks to the origin?

This isn't possible for restricted methods, but is easy for permitted methods.

Reply:
I didn't do it so we can use the same hooks for both, but happy to change if you prefer.

Reply:
We'll have to redefine the hooks regardless since the other hooks are for restricted RPC methods